tpv/core/mapper.py

```diff
@@ -120,8 +121,22 @@ def to_galaxy_destination(self, destination: Destination) -> JobDestination:
 
     def _find_matching_entities(self, tool: GalaxyTool, user: Optional[GalaxyUser]) -> List[EntityWithRules]:
         tool_entity = self.inherit_matching_entities("tools", tool.id)
+
+        # Extract resource requirements from Galaxy tool
+        resource_fields = extract_resource_requirements_from_tool(tool)
+
         if not tool_entity:
-            tool_entity = Tool(evaluator=self.loader, id=tool.id or "unknown_tool_id")
+            # Create new tool entity with resource requirements
+            tool_entity = Tool(evaluator=self.loader, id=tool.id or "unknown_tool_id", **resource_fields)
+        elif resource_fields:
+            # Apply resource requirements to existing tool entity
+            # Create a temporary entity with just the resource fields and combine it
+            resource_tool_entity = Tool(
+                evaluator=self.loader, id=f"{tool.id or "unknown_tool_id"}_resources", **resource_fields
+            )
+            # This seems wrong ? We're probably overriding everything with the (user defined?) resource requirements,
```

Comment:
@nuwang this seems like a conceptual question. In a way the expression provided by a tool should act like some kind of default in the absence of a specific rule for that tool.

Reply:
> if cores and mem are defined in the tpv config for hisat, that should take priority, but if cores and mem are not defined, the resource requirements are applied.

yes, that's right, the issue being that I can't tell what was a default that we should override (or is the default necessarily named default ?)

Reply:
Aah I see - you're referring to resource requirements from the tool wrapper, not user specified resource requests. In that case, it looks like this should be inserted as some kind of base default here as the first element in the matches list: https://github.com/galaxyproject/total-perspective-vortex/blob/25b2709983973d017e9874dc41ef1da69736223a/tpv/core/mapper.py#L49

That way, if a resource requirement spec exists, it'll be overridden by defaults and then by specific matching tools and so on. A bit tricky and looks like it'll need special casing for Tools only.

Reply:
> looks like this should be inserted as some kind of base default here as the first very element in the matches list:

thanks, that's what I was looking for! We should still combine the base with the tool's resource requirements. if we did this I suppose we could pass in the extracted tool requirements ... i'll see where that leads me.